### ISSUE TITLE
Time scale in time domain

### ISSUE THREAD

**Issue by wollmich** [scikit-rf/scikit-rf#860]
I think the following assumption is wrong:

https://github.com/scikit-rf/scikit-rf/blob/cc00e879f615c09a5f84b17355b8362e19c5ae52/skrf/time.py#L410

In the band pass mode the time resolution, maximum time and minimum time are given by:

```
dt_bp = 1 / (n * df)

tp_bp = floor((n - 1) / 2)  * dt_bp
-tn_bp = -floor(n / 2) * dt_bp

tp_bp + tn_bp = (n – 1) / (n * df)
```

In the low pass mode, when the frequency domain data do not include a DC point, the time resolution, maximum time and minimum time are given by:

```
dt_lp = 1 / (2 * n + 1) / df

tp_lp = n * dt_lp
-tn_lp = -n* dt_lp

tp_lp + tn_lp = 2 * n / (2 * n + 1) / df
```

In the low pass mode, when the frequency domain data include a DC point, the time resolution, maximum time and minimum time are given by:

```
dt_lp = 1 / (2 * n - 1) / df

tp_lp = (n - 1) * dt_lp
-tn_lp = -(n - 1) * dt_lp
```

So the anti-alias-time is little less than `1/df`.

**Comment by Vinc0110** [comment]
Could you explain this some more? What exactly is the issue? Did you run into problems with the existing code? Did the time-gating fail in your tests?
You claim the time scale is wrong (it might be), but your list of equations don't seem to proof it. By the way: the time-domain plot created with `Network.plot_s_db_time()` is completely independent from the `time_gate()`.

For further discussion, I propose to treat the low-pass transform without a dc point as a (border line) case of a band-pass transform. Then, we only have to deal with those two cases:

### Band-pass mode:
The frequency-domain signal is treated as a band-pass signal including frequency samples from $f_{min}$ to $f_{max}$. 
For time-gating purposes, the signal is transformed into time-domain using an [inverse FFT](https://numpy.org/doc/stable/reference/generated/numpy.fft.ifft.html), which yields a complex-valued time-domain signal. After gating (multiplication with the time-gate), it gets transformed back into frequency domain using a [FFT](https://numpy.org/doc/stable/reference/generated/numpy.fft.fft.html).
The length of the time-domain signal equals the length of the frequency domain signal: $N_{time} = N_{freq}$.

### Low-pass mode:
The frequency-domain signal is treated as a low-pass signal including frequency samples from $f_{min} = 0$ to $f_{max}$. Mathematically, such a signal actually reaches from $- f_{max}$ to $f_{max}$.
For time-gating purposes, the signal is transformed into time-domain using an inverse FFT, which yields a real-valued time-domain signal ([inverse RFFT](https://numpy.org/doc/stable/reference/generated/numpy.fft.irfft.html)). After gating (multiplication with the time-gate), it gets transformed back into frequency domain using a [RFFT](https://numpy.org/doc/stable/reference/generated/numpy.fft.rfft.html).
The length of the time-domain signal depends on the length of the frequency domain signal:
- $N_{freq}$ is even: $N_{time} = 2 N_{freq} - 1$
- $N_{freq}$ is odd: $N_{time} = 2(N_{freq}-1)$

The odd case is ignored in the code, as I just noticed. That might be another problem and could cause _off-by-one_ errors.

**Comment by Vinc0110** [comment]
And what is "anti-alias-time"?

**Comment by wollmich** [comment]
> And what is "anti-alias-time"?

It's the time span.

**Comment by Vinc0110** [comment]
Please share some more information about the actual problem. Why do you think the time vector is wrong and how does it affect the time gating function? A minimal code example would be great.

**Comment by wollmich** [comment]
> By the way: the time-domain plot created with `Network.plot_s_db_time()` the following property?

Is `Network.plot_s_db_time()` using https://github.com/scikit-rf/scikit-rf/blob/99fb749b4071f6e1cef67fc950dabfa822b5b055/skrf/frequency.py#L635

I think there is the same bug.

You assume there that `t_period = 1/f_step` but I think it should be

```
f_span = (n - 1) * f_step
t_span = (n - 1) * t_step
t_step = 1 / (n * df)
t_span = (n - 1) / (n * f_step)
```

**Comment by wollmich** [comment]
A minimal example:

Let's assume the following network data which represents an offset short (with 10 ps of delay)  in frequency domain:

<HTML>
<BODY>
<TABLE><THEAD><TH>Frequency<br><br>/&#160;GHz</TH><TH>S1,1<br>Zr:&#160;50 Ω<br>Mag</TH><TH>S1,1<br>Zr:&#160;50 Ω<br>Phase&#160;/&#160;&#176;</TH></THEAD><TR><TD>10.000</TD><TD>1.000</TD><TD>108.000</TD></TR><TR><TD>20.000</TD><TD>1.000</TD><TD>36.000</TD></TR><TR><TD>30.000</TD><TD>1.000</TD><TD>-36.000</TD></TR><TR><TD>40.000</TD><TD>1.000</TD><TD>-108.000</TD></TR><TR><TD>50.000</TD><TD>1.000</TD><TD>-180.000</TD></TR></TABLE>

</BODY>
</HTML>

This should give the following data in time domain using the bandpass mode and a rectangular window:

<HTML>
<BODY>
<TABLE><THEAD><TH>Time<br><br>/&#160;ps</TH><TH>TDR&#160;S1,1<br>Zr:&#160;50 Ω<br>Value</TH></THEAD><TR><TD>-40.000</TD><TD>0.000</TD></TR><TR><TD>-20.000</TD><TD>0.000</TD></TR><TR><TD>0.000</TD><TD>0.000</TD></TR><TR><TD>20.000</TD><TD>1.000</TD></TR><TR><TD>40.000</TD><TD>0.000</TD></TR></TABLE>

</BODY>
</HTML>

Note that only the value at 20 ps is exactly 1. All other are 0.


Using your formula `t_period = 1/f_step` the time span would be 100 ps.
Using my formula the `t_span = (n - 1) / (n * f_step)` the time span would be 80 ps.

**Comment by wollmich** [comment]
Using the network data from the above example

[Short_10ps.zip](https://github.com/scikit-rf/scikit-rf/files/11168328/Short_10ps.zip)

your library

```
d = rf.Network('Short_10ps.s1p')
d.plot_s_db_time()
```

returns the following time domain response

![image](https://user-images.githubusercontent.com/22192118/230329856-bdf0a551-edd1-4b30-aab8-111b9f737c1e.png)

where I still think the time axis is wrong. Again here only the point at 20 ps should be 0 dB and others should be -infinity dB. And the time span should be 80 ps and not 100 ps.

Please comment.

**Comment by mhuser** [comment]
You can test with something like:
`d.s11.plot_s_time_step(pad=0)` or `d.s11.plot_z_time_step(pad=0)`
you might need to do [dc extrapolation first](https://scikit-rf.readthedocs.io/en/latest/examples/networktheory/DC%20Extrapolation%20for%20Time%20Domain%20.html)

**Comment by mhuser** [comment]
Let's try this (please mind `plot_s_time_mag` do not share the implementation of `plot_s_time_impulse` and `plot_s_time_response` and is also independent of gating function.
```Python
import skrf as rf
from matplotlib import pyplot as plt
rf.stylely()

n = rf.Network('Short_10ps.s1p');
n_dc = n.extrapolate_to_dc()
window = 'boxcar'
plt.figure()
plt.title('The 10ps short network')
n_dc.plot_s_time_impulse(window = window, pad = 0, label = 'plot_s_time_impulse')
n_dc.plot_s_time_step(window = window, pad = 0, label = 'plot_s_time_step')
n_dc.plot_s_time_mag(label = 'plot_s_time_mag')

plt.savefig('short_10ps.png')
```
![short_10ps](https://user-images.githubusercontent.com/8080934/230471777-6672a94d-cde6-46dc-924a-114b440a821f.png)
To have the time axis being integer multiple of 10 ps, the network must be dc extrapolated (thus 90 ps span because 1 more sample).
However, the time axis of the two group of functions is not consistent, so there is probably an issue.

**Comment by Vinc0110** [comment]
I'm happy to take a look at this in the next couple of days.

With the 10ps delay example, I don't really get the point. Why should a time span of 100 ps be wrong and 80 ps be correct?

In any case, I'm wondering if there is a known correct reference to compare against the scikit-rf implementation and @wollmich's equations. So far, I think both methods are not at all supported by theory. We should improve that.

**Comment by wollmich** [comment]
> With the 10ps delay example, I don't really get the point. Why should a time span of 100 ps be wrong and 80 ps be correct?

If we have `n = 5` frequency points and a frequency step of `f_step = 10 GHz` the time span will be `t_span = (n - 1) / (n * f_step) = 80 ps`. This time signal will be period with `t_period = n / (n * f_step) = 1 / f_step = 100 ps`.

**Comment by FranzForstmayr** [comment]
@wollmich you are right and triggered more than one issue at once. Thanks for the report!

`plot_s_db_time` applies a window function by default (kaiser, 6). With that small number of samples there's an significant error compared to `boxcar` in your numpy implementation. #900 let's you specify a custom window and you can override this default behaviour.

The time displayed for this plot was taken from `Frequency.t`. The current implementation uses a wrong sampling time.

You provided an example with 50GHz bandwith, which would need a sampling frequency of 100 GSps which would result in a sampling time of 10 ps. You are using a step of 20 ps and currently #900 does the same, but I think it's safe to use 10 ps. The PR is still WIP, using 10 ps causes to fail several other tests right now.


@Vinc0110 @mhuser 
Look at this example code:
```python
# Current behaviour
f = rf.Frequency(10, 50, 5)
f.t 
array([-5.0e-11, -2.5e-11,  0.0e+00,  2.5e-11,  5.0e-11]) # 25 ps would imply 40 GSps = 20GHz

# PR 900 
f = rf.Frequency(10, 50, 5)
f.t 
array([-4.0e-11, -2.0e-11,  0.0e+00,  2.0e-11,  4.0e-11]) # 20 ps would imply 50 GSps = 25 GHz

# I would like to have when PR900 is finished
f = rf.Frequency(10, 50, 5)
f.t 
array([-5.0e-11, -4.0e-11, -3.0e-11, -2.0e-11, -1.0e-11 0.0e+00,  1.0e-11, 2.0e-11, 3.0e-11, 4.0e-11]) # 10 ps would imply 100 GSps = 100 GHz
```

**Comment by FranzForstmayr** [comment]
Should be fixed in #900
